utils/src/main/java/software/amazon/awssdk/utils/CollectionUtils.java

```diff
@@ -139,4 +140,26 @@ public static boolean isNullOrEmpty(Map<?, ?> map) {
     public static <K, V> Map<K, V> inverseMap(Map<V, K> inputMap) {
         return inputMap.entrySet().stream().collect(Collectors.toMap(Map.Entry::getValue, Map.Entry::getKey));
     }
+
+    /**
+     * For a collection of values of type {@code V} that can all be converted to type {@code K}, create a map that
+     * indexes all of the values by {@code K}. This requires that no two values map to the same index.
+     * 
+     * @param values the collection of values to index
+     * @param indexFunction the function used to convert a value to its index
+     * @param <K> the index (or key) type
+     * @param <V> the value type
+     * @return a (modifiable) map that indexes K to its unique value V
+     * @throws IllegalArgumentException if any of the values map to the same index
+     */
+    public static <K, V> Map<K, V> index(Iterable<V> values, Function<? super V, K> indexFunction) {
+        Map<K, V> map = new HashMap<>();
+        for (V value : values) {
+            K index = indexFunction.apply(value);
+            V prev = map.put(index, value);
+            Validate.isNull(prev, "No duplicate indices allowed but both %s and %s have the same index: %s",
```

Comment:
Query : Since we are indexing to unique Converted type should that be reflected in the function name ?

In some case in future I might even want to index duplicate value in where I would expect  Map<K, List<V>> might have difficulty to use the same function name in this class

Reply:
Good point/question.

My thinking is that indexing like this would be the common/standard use case, where users expect unique indices, and if we wanted to extend this to something like multi-map support, then that new function name would be specialized to differentiate. E.g., `index` vs `indexMulti`, or similar.

However, since the current function can throw upon encounter duplicates, then perhaps there is an argument for giving this one more explicit naming. What do you think?

Reply:
Renamed `index` to `uniqueIndex`.